### PR TITLE
External OpenXR Loader support

### PR DESCRIPTION
**Pull request by kisg**
Add support for using an external OpenXR loader library, instead of the built-in KHR loader.

This is required on the Meta Oculus Quest devices, where Meta uses a custom OpenXR
loader library. To use it, specify the loader library as a build parameter, e.g.:
scons platform=android android_arch=arm64v8 external_openxr_library=/path/to/libopenxr_loader.so

The change also adds support for building the KHR loader for Android, if an external loader
is not requested.

Tested with the Godot 4 OpenXR Demo on a Meta Quest 2.

### CODE REVIEW

**Review by m4gr3d — COMMENTED** [review]
@kisg What are the files under `thirdparty/openxr/src/external/android-jni-wrappers` added for?

**Comment by m4gr3d** [comment]
> To use it, specify the loader library as a build parameter, e.g.:
scons platform=android android_arch=armv8 external_openxr_library=/path/to/libopenxr_loader.so

@kisg How would we expose the capability to regular end users?

**Comment by kisg** [comment]
> @kisg What are the files under `thirdparty/openxr/src/external/android-jni-wrappers` added for?

The PR also enables the building of the KHR loader for Android (I will extend the PR / commit message with this), and these JNI wrappers are required for that.

**Comment by kisg** [comment]
> > To use it, specify the loader library as a build parameter, e.g.:
> > scons platform=android android_arch=armv8 external_openxr_library=/path/to/libopenxr_loader.so
> 
> @kisg How would we expose the capability to regular end users?

Currently, if you rebuild the engine with an external OpenXR loader  library, and then generate the Android template (./gradlew generateGodotTemplates), the external loader library will be included in the template, so you may use it with the Custom Build Template option during export.

While this is not ideal, it is definitely a step forward from the current state, where it is not possible to target the Quest devices.

I can imagine a follow-up PR, that could make things even more straightforward, especially if it would be accepted by the Godot maintainers to store the Meta OpenXR loader in the repository.

In this case:
- the KHR loader could also be built as a shared library (only for Android)
- the Meta and the KHR loaders would be included in the Android Gradle templates as separate build variants / product flavors.
- during the export process the desired build variant could be chosen.

What do you think?

**Comment by BastiaanOlij** [comment]
Thanks @kisg , I'll give this a proper review soon but at first glance this is along the lines of what I wanted to get done myself.

The default builds should be with the Khronos loader, switching in the Meta or Pico loader for those vendors as an option, where people can supply the external loaders themselves and custom build.
We can do "unofficial" nightly builds or something in the line of what Calino does for nightly dev builds. That allows us to get around our mandate of not including closed source binaries. 

That said, still hope Meta will switch over to the Khronos loader.

**Comment by m4gr3d** [comment]
> We can do "unofficial" nightly builds or something in the line of what Calino does for nightly dev builds. That allows us to get around our mandate of not including closed source binaries.

@BastiaanOlij I think what @kisg is referring to is to generate regular android build templates and 'meta' android build templates each including the corresponding loader, and update the export logic so it knows which to select based on the user's export preferences.

So the end user still have the option easily available without requiring them to know how to generate a custom build.

There's still an issue of the question of how to automate this process, since that's how the build templates are generated, but this could be circumvented by pushing the meta loader binaries into its own repo, and having the CI build process checks it out at build time.

**Comment by BastiaanOlij** [comment]
> There's still an issue of the question of how to automate this process, since that's how the build templates are generated, but this could be circumvented by pushing the meta loader binaries into its own repo, and having the CI build process checks it out at build time.

The problem remains, as I understand it, that we can't make this part of the official build system. Having a dependency on closed source binaries in anything we produce is a no-go. We can include the optional logic if the person doing the build provides the closed source part but that means a custom build.

This is one of the main reasons why the Godot 3 plugins are not in the main godotengine repo but in the GodotVR repo. 

For Godot 4 it likely means automating the build on a fork or maybe asking Calinou to add it to his nightly build scripts and making a quest-android template available as a 3rd party download.

Or hopefully one day soon Meta will support the Khronos loader... ;)

File: modules/openxr/SCsub
Comment (BastiaanOlij) [inline modules/openxr/SCsub:50]: It's not entirely clear to me why `env_thirdparty` and `env_khrloader` are split here. The idea in the original SCsub is that `env_thirdparty` contains all thirdparty sources while `env_openxr` contains our modules sources. The KHR loader is part of the third party sources.

File: modules/openxr/SCsub
Comment (BastiaanOlij) [inline modules/openxr/SCsub:107]: I was looking that this before and I think a nicer solution is to move the original check that adds the defines  to `env_openxr` before `env_thirdparty` is cloned from it so we have consistent settings here.

**Comment by m4gr3d** [comment]
> This is one of the main reasons why the Godot 3 plugins are not in the main godotengine repo but in the GodotVR repo.
>
> For Godot 4 it likely means automating the build on a fork or maybe asking Calinou to add it to his nightly build scripts and making a quest-android template available as a 3rd party download.

Actually you inadvertently suggested a path to make it work :)

The GodotVR repo (or another repo like it) can be used to generate the quest-android template in a similar manner it does for Godot 3. The difference being instead of generating an AAR plugin, it generates the entire Android build template.

We would then need to instruct the user to get the quest-android template and install it manually (the option is already included in the editor) to replace the default android template.

I think it's worth investigating given that there may be other vendors that are also slow at adopting khronos's standards.

**Comment by BastiaanOlij** [comment]
> > This is one of the main reasons why the Godot 3 plugins are not in the main godotengine repo but in the GodotVR repo.
> > For Godot 4 it likely means automating the build on a fork or maybe asking Calinou to add it to his nightly build scripts and making a quest-android template available as a 3rd party download.
> 
> Actually you inadvertently suggested a path to make it work :)
> 
> The GodotVR repo (or another repo like it) can be used to generate the quest-android template in a similar manner it does for Godot 3. The difference being instead of generating an AAR plugin, it generates the entire Android build template.
> 
> We would then need to instruct the user to get the quest-android template and install it manually (the option is already included in the editor) to replace the default android template.
> 
> I think it's worth investigating given that there may be other vendors that are also slow at adopting khronos's standards.

Owh yes ! this is exactly what I'm getting at :)

**Comment by akien-mga** [comment]
Can't we compile support for the Meta loader without the binary lib, and `dlopen` it if provided, otherwise fallback to the Khronos loader?

Providing a whole set of templates just for Meta Quest sounds like a hassle. But we could have an easy way for people to download the Meta loader to their project files for use in Custom Build.

**Comment by BastiaanOlij** [comment]
> Can't we compile support for the Meta loader without the binary lib, and `dlopen` it if provided, otherwise fallback to the Khronos loader?
> 
> Providing a whole set of templates just for Meta Quest sounds like a hassle. But we could have an easy way for people to download the Meta loader to their project files for use in Custom Build.

No idea, I think its a static lib we're linking in, not a dynamic one so I guess no. Plus that Pico for some reason decided to do their own loader as well, so we would have to start doing this trick for each vendor that decides not to follow the standard. 

I fully understand Meta here as they had to create a loader before Khronos had finalized the standard on Android, but why Pico didn't just adopt the full standard from day one... Both Monado and HTC have adopted the standard properly.

In the end though, this is a temporary situation that will hopefully resolve itself soon, so being able to switch which loader is compiled in and providing a Quest centric template to me seems like the least compromising solution. I'm pretty positive that we'll see Meta adopting the Khronos loader in due time.

File: modules/openxr/SCsub
Comment (kisg) [reply modules/openxr/SCsub]: Should be done in the latest version of the PR.

File: modules/openxr/SCsub
Comment (kisg) [reply modules/openxr/SCsub]: Moved as requested.

File: modules/openxr/config.py
Comment (kisg) [reply modules/openxr/config.py]: Done in latest updates.

**Comment by BastiaanOlij** [comment]
@akien-mga I'm actually keen to approve this, possibly we should handle this PR in our next PR meeting so others can way in if we're ok with this approach? I'll try and be present for that one.

File: modules/openxr/SCsub
Comment (BastiaanOlij) [inline modules/openxr/SCsub:25]: So what I meant here was moving the whole 
```
if env["platform"] == "android":
    # may need to set OPENXR_ANDROID_VERSION_SUFFIX
    env_thirdparty.AppendUnique(CPPDEFINES=["XR_OS_ANDROID", "XR_USE_PLATFORM_ANDROID"])

    # may need to include java parts of the openxr loader
elif env["platform"] == "linuxbsd":
    env_thirdparty.AppendUnique(CPPDEFINES=["XR_OS_LINUX"])

    if env["x11"]:
        env_thirdparty.AppendUnique(CPPDEFINES=["XR_USE_PLATFORM_XLIB"])

    # FIXME: Review what needs to be set for Android and macOS.
    env_thirdparty.AppendUnique(CPPDEFINES=["HAVE_SECURE_GETENV"])
elif env["platform"] == "windows":
    env_thirdparty.AppendUnique(CPPDEFINES=["XR_OS_WINDOWS", "NOMINMAX", "XR_USE_PLATFORM_WIN32"])
```
Block above `env_thirdparty = env_openxr.Clone()` and applying it to `env_openxr`.

This will mean that the correct defines are also in use when we include our openxr headers into our module, I suspect running into this eventually for WMR support. 

It also prevents adding the same defines twice.

Comment (kisg) [reply modules/openxr/SCsub:25]: Updated as requested. Could you please check it?

**Comment by kisg** [comment]
> @akien-mga I'm actually keen to approve this, possibly we should handle this PR in our next PR meeting so others can way in if we're ok with this approach? I'll try and be present for that one.

@BastiaanOlij @akien-mga 
We managed to implement a better approach, inspired by the discussion here.

It looks like the ABI of our KHR loader library and Meta's OpenXR loader library is the same. I think that Pico's will be the same as well, but I could not check it yet.

This means that if we build our loader as a shared library, and link it with libgodot_android.so, and then during export we replace the original libopenxr_loader.so with Meta's / Pico's OpenXR loader, it will just work. We actually managed to get this running just now on our Quest 2.

The next step that we are working on now is to package these external OpenXR Loader libraries as Godot Android Plugins.

This would allow for a much better Godot user story:
-  We just need 1 Android template, that will include the KHR OpenXR loader.
- The device specific loaders can be distributed as Android plugins in the Asset store.

There is only one limitation with this approach: due to the limitations of the current Godot Android build system, only 1 of such OpenXR Loader plugin can be added to a project at the same time. However, it should be possible to lift this limitation in the future by leveraging the product flavor functionality in the Android Gradle Plugin and extending the Godot Android build accordingly, so it will be possible to build for different devices (product flavors) from the same source tree with multiple such plugins present.

We will submit this new implementation as soon as it is ready (possible as a separate PR?).

What do you think?

**Comment by akien-mga** [comment]
That sounds great to me :+1:

**Comment by BastiaanOlij** [comment]
Sounds amazing!! Reading that though I find it even more mysterious why Meta hasn't just switched to the Khronos loader. All the damn thing does is get pointers to function calls into OpenXR..

**Comment by BastiaanOlij** [comment]
btw @kisg , would it make sense to put the changes for the Khronos loader in a separate PR we merge first so that we have the Android logic completed, and then do a PR with the additions allowing us to swap in the Meta or PICO loaders?

**Comment by kisg** [comment]
@BastiaanOlij Sure, that is an option as well, I just wanted to add the necessary Gradle changes (due to the Android Gradle Plugin's limitations we need to patch the APK at the end of the build to add the default libopenxr_loader.so if there was not one added by a plugin). I hope to have it all done by Monday. If not, I will send a separate PR with just the core changes.

**Comment by m4gr3d** [comment]
> @BastiaanOlij Sure, that is an option as well, I just wanted to add the necessary Gradle changes (due to the Android Gradle Plugin's limitations we need to patch the APK at the end of the build to add the default libopenxr_loader.so if there was not one added by a plugin). I hope to have it all done by Monday. If not, I will send a separate PR with just the core changes.

@kisg from your description, the changes should be separated in separate PRs since one is XR related and the other touches the Android export logic. 
It'll also make it easier to review and address concerns specific to each PR.

**Comment by kisg** [comment]
A small update from our side: we implemented the solution that we talked about here, with dynamically linking libopenxr_loader.so to libgodot_android.so directly, but I did not like that we had to do ugly patching at the end of the Android build: essentially checking if there was the loader library in the final APK, and adding the KHR one if it was missing.

So we changed the OpenXR module (class OpenXRAPI to be exact) to load the openxr_loader dynamically at runtime (using OS::open_dynamic_library) and resolve all symbols with xrGetInstanceProcAddr.

This has the added benefit that (according to the [Khronos docs](https://registry.khronos.org/OpenXR/specs/1.0/loader.html)) this reduces the number of indirections when calling OpenXR functions because we store the function pointers into the OpenXR runtime directly and don't have to go through the trampoline table of the loader.

We will post this new version as a separate PR, as soon as we have cleaned it up.

I am also thinking, that with this new solution we would not need to include the OpenXR loader code in the main binary on any platform. My assumption is that on any system where an OpenXR runtime is available, there will be a suitable OpenXR loader library at a "well-known" location that we can just open at runtime with dlopen(). This could reduce the size of the main Godot binary by ~500K + reduce the source code size as well, because we would only need to keep the headers in the source.

Android based headsets seem to be an exception to this rule, but there we will provide separate Android plugins with the suitable loaders (from the GodotVR / XR project) anyways.

What do you think @BastiaanOlij @m4gr3d @akien-mga?

**Comment by BastiaanOlij** [comment]
Hey @kisg,

I've been wondering about replacing the Khronos loader with something more lightweight but are we going to be feature compatible on desktop? On desktop the loader does a whole bunch more in figuring out what the active XR runtime is, with environment variable overrule, and then loading the correct loader?

Seeing that we're already obtaining pointers for most of the extensions getting pointers for some of the main functions instead of relying on the code in the Khronos loader isn't that big a step.

**Comment by BastiaanOlij** [comment]
> My assumption is that on any system where an OpenXR runtime is available, there will be a suitable OpenXR loader library at a "well-known" location that we can just open at runtime with dlopen()

Actually, this assumption is wrong. On Linux and Windows there are different approaches in place. I'm not 100% sure about the Linux one but on Windows you need to either check an environment variable that points to a JSON configuration file or check the registry for the current active XR runtime, then load the JSON and parse it, and it tells you which DLL to load.

Linux also uses the JSON files but the way you determine which one to load is different.

It wouldn't be super difficult to re-produce the logic but it's not as simple as pointing to a fixed location. Every XR runtime has complete freedom in installing the relevant files and often the user has control over that installation folder.

**Comment by BastiaanOlij** [comment]
There is no loader library on Windows as a system library, only the XR runtime. The loader DLL can be placed alongside the exe but we chose to statically link it seeing that is how we always deploy Godot. We could offcourse keep it statically linked on Windows and Linux and only do the dynamic loading on Android

**Comment by kisg** [comment]
> There is no loader library on Windows as a system library, only the XR runtime. The loader DLL can be placed alongside the exe but we chose to statically link it seeing that is how we always deploy Godot. We could offcourse keep it statically linked on Windows and Linux and only do the dynamic loading on Android

Ok, we will do it like that.

**Comment by BastiaanOlij** [comment]
Btw I'm planning on updating the thirdparty library to 1.0.25 now that has been released, is that going to clash with what you're doing?

**Comment by kisg** [comment]
> Btw I'm planning on updating the thirdparty library to 1.0.25 now that has been released, is that going to clash with what you're doing?

No, that should not be a problem. We just need the dependencies for the android build as well, but we can add them if necessary.

File: modules/openxr/SCsub
Comment (BastiaanOlij) [reply modules/openxr/SCsub:25]: Yup, perfect!

**Review by BastiaanOlij — APPROVED** [review]
I'm going to tentatively approve this PR. I do think we need a bit more discussion before merging however as a first stab at getting full support working on Android for devices that support the Khronos loader, and a way for people to manually build a Quest runtime, this is a major step forward.

We can enhance it from here to see if we can get a better solution to make it easier for people to deploy on Quest or Pico. Hopefully other vendors will simply jump on board the Khronos bandwagon.

Couple of nitpicky things:
- I'm assuming the thirdparty files have been updated to 1.0.25
- thirdparty/readme should be edited so it properly informs the need to include the android platform files
- Should probably squash the commits to one commit.

**Comment by kisg** [comment]
Here is the new version of the PR with dynamic loading of OpenXR loader library on Android: https://github.com/godotengine/godot/pull/65798

**Comment by kisg** [comment]
Superseded by #65798.

**Pull request closed without merging**
